### ISSUE TITLE
Update tree stats API for use numpy natively

### ISSUE THREAD

**Issue by jeromekelleher** [jeromekelleher/msprime#394]
#391 changed the TreeSequence.samples method to return a numpy array, and we should aim to make all APIs that interact with sets of samples work efficiently with numpy arrays of node IDs. The tree stats API currently uses a lot of list operations, and we should update it to use numpy fully.

cc @petrelharp.

**Comment by petrelharp** [comment]
Agree, but the list operations are a minor contribution to run time compared to the actual calculations.

**Comment by jeromekelleher** [comment]
> Agree, but the list operations are a minor contribution to run time compared to the actual calculations.

That's probably true. The main motivation is that it makes it much simpler to communicate with C code when you use numpy arrays. No parsing Python lists in C (which is super tedious).
